decompiler/frontend/binaryninja/handlers/types.py

```diff
@@ -75,30 +75,46 @@ def lift_named_type_reference_type(self, custom: NamedTypeReferenceType, **kwarg
 
     def lift_enum(self, binja_enum: EnumerationType, name: str = None, **kwargs) -> Enum:
         """Lift enum type."""
+        type_id = hash(binja_enum)
         enum_name = name if name else self._get_data_type_name(binja_enum, keyword="enum")
+        if enum_name.strip() == "":
+            enum_name = f"__anonymous_enum"
+        enum_name = self._lifter.unique_name_provider.get_unique_name(enum_name)
         enum = Enum(binja_enum.width * self.BYTE_SIZE, enum_name, {})
         for member in binja_enum.members:
             enum.add_member(self._lifter.lift(member))
-        self._lifter.complex_types.add(enum)
+        self._lifter.complex_types.add(enum, type_id)
         return enum
 
     def lift_enum_member(self, enum_member: EnumerationMember, **kwargs) -> ComplexTypeMember:
         """Lift enum member type."""
         return ComplexTypeMember(size=0, name=enum_member.name, offset=-1, type=Integer(32), value=int(enum_member.value))
 
-    def lift_struct(self, struct: StructureType, name: str = None, **kwargs) -> Union[Struct, ComplexTypeName]:
+    def lift_struct(self, struct: StructureType, name: str = None, **kwargs) -> Union[Struct, Union_, Class, ComplexTypeName]:
+        type_id = hash(struct)
+        cached_type = self._lifter.complex_types.retrieve_by_id(type_id)
+        if cached_type is not None:
+            return cached_type
+
         """Lift struct or union type."""
         if struct.type == StructureVariant.StructStructureType:
-            type_name = name if name else self._get_data_type_name(struct, keyword="struct")
-            lifted_struct = Struct(struct.width * self.BYTE_SIZE, type_name, {})
+            keyword, type, members = "struct", Struct, {}
         elif struct.type == StructureVariant.UnionStructureType:
-            type_name = name if name else self._get_data_type_name(struct, keyword="union")
-            lifted_struct = Union_(struct.width * self.BYTE_SIZE, type_name, [])
+            keyword, type, members = "union", Union_, []
+        elif struct.type == StructureVariant.ClassStructureType:
+            keyword, type, members = "class", Class, {}
         else:
             raise RuntimeError(f"Unknown struct type {struct.type.name}")
+
+        type_name = name if name else self._get_data_type_name(struct, keyword=keyword)
+        if type_name.strip() == "":
```

Comment:
Lines 110-112 looks almost identical to 80-82. Do you think it makes sense to move them to the separate method?

Reply:
Yes, sounds good! I could add this functionality into the existing `_get_data_type_name`.

Reply:
fixed by [9036c07](https://github.com/fkie-cad/dewolf/pull/353/commits/9036c07e9af6964b8209c31ab197407203d7a2a3)